pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -627,6 +634,338 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+			og.recorder.Eventf(volumeToMount.Pod, v1.EventTypeWarning, kevents.FailedMapVolume, eventErr.Error())
+			return detailedErr
+		}
+
+		// Take filedescriptor lock to keep a block device opened. Otherwise, there is a case
+		// that the block device is silently removed and attached another device with same name.
+		// Container runtime can't handler this problem. To avoid unexpected condition fd lock
+		// for the block device is required.
+		_, err = og.blkUtil.AttachFileDevice(devicePath)
+		if err != nil {
+			return volumeToMount.GenerateErrorDetailed("MapVolume.AttachFileDevice failed", err)
+		}
+
+		// Device mapping for pod device map path succeeded
+		simpleMsg, detailedMsg = volumeToMount.GenerateMsg("MapVolume.MapDevice succeeded", fmt.Sprintf("volumeMapPath %q", volumeMapPath))
+		verbosity = glog.Level(1)
+		og.recorder.Eventf(volumeToMount.Pod, v1.EventTypeNormal, kevents.SuccessfulMountVolume, simpleMsg)
+		glog.V(verbosity).Infof(detailedMsg)
+
+		// Update actual state of world
+		markVolMountedErr := actualStateOfWorld.MarkVolumeAsMounted(
+			volumeToMount.PodName,
+			volumeToMount.Pod.UID,
+			volumeToMount.VolumeName,
+			nil,
+			blockVolumeMapper,
+			volumeToMount.OuterVolumeSpecName,
+			volumeToMount.VolumeGidValue)
+		if markVolMountedErr != nil {
+			// On failure, return error. Caller will log and retry.
+			return volumeToMount.GenerateErrorDetailed("MapVolume.MarkVolumeAsMounted failed", markVolMountedErr)
+		}
+
+		return nil
+	}, blockVolumePlugin.GetPluginName(), nil
+}
+
+// GenerateUnmapVolumeFunc marks volume as unmonuted based on following steps.
+// Remove symbolic links from global map path and pod device map path.
+// Once those cleanups are done, remove pod device map path dir.
+// If all steps are completed, the volume is marked as unmounted.
+func (og *operationGenerator) GenerateUnmapVolumeFunc(
+	volumeToUnmount MountedVolume,
+	actualStateOfWorld ActualStateOfWorldMounterUpdater) (func() error, string, error) {
+
+	// Get block volume unmapper plugin
+	var blockVolumeUnmapper volume.BlockVolumeUnmapper
+	blockVolumePlugin, err :=
+		og.volumePluginMgr.FindMapperPluginByName(volumeToUnmount.PluginName)
+	if err != nil {
+		return nil, "", volumeToUnmount.GenerateErrorDetailed("UnmapVolume.FindMapperPluginByName failed", err)
+	}
+	if blockVolumePlugin == nil {
+		return nil, "", volumeToUnmount.GenerateErrorDetailed("UnmapVolume.FindMapperPluginByName failed to find BlockVolumeMapper plugin. Volume plugin is nil.", nil)
+	}
+	blockVolumeUnmapper, newUnmapperErr := blockVolumePlugin.NewBlockVolumeUnmapper(
+		volumeToUnmount.InnerVolumeSpecName, volumeToUnmount.PodUID)
+	if newUnmapperErr != nil {
+		return nil, blockVolumePlugin.GetPluginName(), volumeToUnmount.GenerateErrorDetailed("UnmapVolume.NewUnmapper failed", newUnmapperErr)
+	}
+
+	return func() error {
+		// Try to unmap symlink on global map path
```

Comment:
Should it be unmap pod link first, then global map?

Reply:
yeah, it's reasonable.